### PR TITLE
Add ID token verification helpers.

### LINKED ISSUE #39: Implement google.oauth2.verify_id_token
(Towards #3)

### PR DESCRIPTION
**Pull request by jonparrott**
Resolves #39 and the v1.0.0 milestone.

### CODE REVIEW

File: google/oauth2/id_token.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+_GOOGLE_OAUTH2_CERTS_URL = 'https://www.googleapis.com/oauth2/v1/certs'
+
+# The URL that provides public certificates for verifying ID tokens issued
+# by Firebase and the Google APIs infrastructure
+_GOOGLE_APIS_CERTS_URL = (
+    'https://www.googleapis.com/robot/v1/metadata/x509'
+    '/securetoken@system.gserviceaccount.com')
+
+
+def _fetch_certs(request, certs_url):
```
Comment (dhermes) [inline google/oauth2/id_token.py]: 1. Needs docstring
1. Don't you want some caching for this?

File: google/oauth2/id_token.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    response = request('GET', certs_url)
+
+    if response.status != http_client.OK:
+        raise exceptions.TransportError(
+            'Could not fetch certificates at {}'.format(certs_url))
+
+    return json.loads(response.data.decode('utf-8'))
```
Comment (dhermes) [inline google/oauth2/id_token.py]: 1. Do we care about JSON parsing failure here?
1. Do we have a guarantee that all transports have `bytes` as `data`?

File: google/oauth2/id_token.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+def verify_token(id_token, request, audience=None,
+                 certs_url=_GOOGLE_OAUTH2_CERTS_URL):
+    """Verifies an ID token and returns the decoded token.
+
+    Args:
+        id_token (Union[str, bytes]): The encoded token.
+        request (google.auth.transport.Request): The object used to make
+            HTTP requests.
+        audience (str): The audience that this token is intended for. If None
```
Comment (dhermes) [inline google/oauth2/id_token.py]: `Optional[str]`

File: google/oauth2/id_token.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            then the audience is not verified.
+        certs_url (str): The URL that specifies the certificates to use to
```
Comment (dhermes) [inline google/oauth2/id_token.py]: `Optional[str]`

File: google/oauth2/id_token.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+
+def verify_oauth2_token(id_token, request, audience=None):
+    """Verifies an ID Token issued by Google's OAuth 2.0 authorization server.
+
+    Args:
+        id_token (Union[str, bytes]): The encoded token.
+        request (google.auth.transport.Request): The object used to make
+            HTTP requests.
+        audience (str): The audience that this token is intended for. This is
```
Comment (dhermes) [inline google/oauth2/id_token.py]: `Optional[str]`

File: google/oauth2/id_token.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+            then the audience is not verified.
+        certs_url (str): The URL that specifies the certificates to use to
+            verify the token. This URL should return JSON in the format of
+            ``{'key id': 'x509 certificate'}``.
+
+    Returns:
+        Mapping[str, Any]: The decoded token.
+
+    Raises:
+        ValueError: if verification fails.
```
Comment (dhermes) [inline google/oauth2/id_token.py]: This error isn't raised here, typically `Raises` refers to explicitly raised exceptions

File: google/oauth2/id_token.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+            HTTP requests.
+        audience (str): The audience that this token is intended for. This is
+            typically your application's OAuth 2.0 client ID. If None then the
+            audience is not verified.
+
+    Returns:
+        Mapping[str, Any]: The decoded token.
+
+    Raises:
+        ValueError: if verification fails.
```
Comment (dhermes) [inline google/oauth2/id_token.py]: Ditto on Raises section

File: google/oauth2/id_token.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    """
+    return verify_token(
+        id_token, request, audience=audience,
+        certs_url=_GOOGLE_OAUTH2_CERTS_URL)
+
+
+def verify_firebase_token(id_token, request, audience=None):
```
Comment (dhermes) [inline google/oauth2/id_token.py]: Whoa! On fire!

File: tests/oauth2/test_id_token.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+from google.auth import exceptions
+from google.oauth2 import id_token
+
+
+def make_request(status, data=None):
+    response = mock.Mock()
+    response.status = status
+
+    if data is not None:
```
Comment (dhermes) [inline tests/oauth2/test_id_token.py]: You remembered your nunchucks!

File: google/oauth2/id_token.py (hunk 1 above)
Comment (jonparrott) [inline google/oauth2/id_token.py]: 1. Done.
2. It's up to the transport. Requests has requests-cache.

File: google/oauth2/id_token.py (hunk 2 above)
Comment (jonparrott) [inline google/oauth2/id_token.py]: 1. I checked in other places where we do this (like in _client) and in general we don't explicitly deal with JSON parse errors. There's a bit of an expectation that if response code is 200 that the data is also valid. Anything else is "truly exceptional". The failure cases, such as when dealing with 500s, explicitly handles bad JSON.
2. Yes.

File: google/oauth2/id_token.py (hunk 3 above)
Comment (jonparrott) [inline google/oauth2/id_token.py]: > An optional type is also automatically assumed when the default value is None

https://www.python.org/dev/peps/pep-0484/#union-types

File: google/oauth2/id_token.py (hunk 6 above)
Comment (jonparrott) [inline google/oauth2/id_token.py]: Fair enough.

File: google/oauth2/id_token.py (hunk 8 above)
Comment (jonparrott) [inline google/oauth2/id_token.py]: 🔥 🔥 🔥

File: tests/oauth2/test_id_token.py (hunk 9 above)
Comment (jonparrott) [inline tests/oauth2/test_id_token.py]: I did!

File: google/oauth2/id_token.py (hunk 3 above)
Comment (dhermes) [inline google/oauth2/id_token.py]: I was strictly pointing it out for usability of docs

**Pull request merged**